_episodes/08-exercise.md

```diff
@@ -15,6 +17,87 @@ Now we have developed a range of skills relating to testing and continuous
+maintain) confidence into the correctness of your code. But in detail your tests
+can serve a variety of purposes. It can be useful to keep in mind what you could
+use tests for during your coding, so we compiled a certainly non-exhaustive list
+of test purposes here. The major ones were discussed in the lesson; some more
+exotic ones should be seen as suggestions for you to try. A test can have more
+than one of these purposes:
+
+* **test for features** - The first and simplest test that is capable of verifying
+  the correctness of any feature (in the broadest sense) of your code. This is
+  the obvious purpose of a test and encountered everywhere in the lesson.
+* **test for confidence (in a narrow sense)** - Additional tests of the same
+  features that are redundant in that they repeat a test for features with
+  qualitatively similar input just to double-check. This is also encountered in
+  the lesson, e.g. in the solution of "More parameters" in [pytest
+  features][pytest_features] large numbers aren't really that different from
+  other numbers unless you run into overflow errors (one could even argue that
+  testing the negative numbers in [pytest features][pytest_features] is not
+  qualitatively different and rather  to double check). One should be aware of
+  the difference between testing for confidence and necessary feature tests. The
+  former is a convenience that comes at the cost of longer test runs and so it
+  is not always desirable to test redundantly (although certainly better than
+  missing an aspect).
+* **test for edge-/corner-cases** - Test special input or conditions for which a
+  general algorithm needs to be specialised (e.g., NaNs, infinities, overflows,
+  empty input, etc.). We did a [whole episode][edge_cases] on this.
+* **test for failures** - This is part of feature testing but important enough to
+  mention explicitly: The conditions under which your code fails are part of
+  your interface and need to be tested. The user (that probably includes
+  yourself) might rely on a raised exception or returned default value in the
+  case of failure. Make sure they can and think of all the cases that your
+  current approach cannot handle. Any changes in these (even those for the
+  better) are changes of the interface and should appear intentionally. This was
+  discussed in [pytest features][pytest_features].
+* **fuzzy testing** - This is broader than testing for failures; if you have
+  unexperienced or even malicious users of your project, they might run your
+  code with inputs or under conditions that do not make any sense at all and are
+  almost impossible to predict. Fuzzy testing is a strategy where you let the
+  computer run your code with random input (sometimes down to the bit level) and
+  make sure that not even the most far-fetched input can break your code. There
+  are libraries for that, so you don't have to set up all the boilerplate
```

Comment:
Worth explicitly mentioning one or two such libraries here?

Reply:
Haven't done fuzzy testing myself, yet, so I neither have opinions nor experience. I know about Hypothesis but I guess the any link I can provide is about as reliable as their own answer to hacking "fuzzy testing python" into their preferred search engine. If you have suggestions, I'd say you add that after the pull request is through, okay?

Reply:
I've used Hypothesis a couple of times and it's good; I've not worked with other kinds of fuzzer though so have no suggestions on that front. Happy to add a bullet after merging.